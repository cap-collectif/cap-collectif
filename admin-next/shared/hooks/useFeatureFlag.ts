import { useLazyLoadQuery, graphql } from 'react-relay'
import invariant from '../utils/invariant'
import { useFeatureFlagQuery, FeatureFlagType } from '@relay/useFeatureFlagQuery.graphql'

type FeatureFlags = Omit<
  {
    [key in FeatureFlagType]: boolean
  },
  '%future added value'
>

const query = graphql`
  query useFeatureFlagQuery {
    featureFlags {
      type
      enabled
    }
  }
`
export const defaultFeatureFlags: FeatureFlags = {
  report_browers_errors_to_sentry: false,
  login_saml: false,
  oauth2_switch_user: false,
  votes_min: false,
  blog: false,
  calendar: false,
  login_facebook: false,
  privacy_policy: false,
  members_list: false,
  captcha: false,
  turnstile_captcha: false,
  questionnaire_result: false,
  consent_external_communication: false,
  consent_internal_communication: false,
  newsletter: false,
  profiles: false,
  projects_form: false,
  project_trash: false,
  search: false,
  share_buttons: false,
  shield_mode: false,
  registration: false,
  phone_confirmation: false,
  reporting: false,
  themes: false,
  districts: false,
  user_type: false,
  votes_evolution: false,
  restrict_registration_via_email_domain: false,
  export: false,
  server_side_rendering: false,
  zipcode_at_register: false,
  consultation_plan: false,
  display_map: false,
  sso_by_pass_auth: false,
  allow_users_to_propose_events: false,
  secure_password: false,
  restrict_connection: false,
  login_franceconnect: false,
  read_more: false,
  display_pictures_in_depository_proposals_list: false,
  external_project: false,
  multilangue: false,
  display_pictures_in_event_list: false,
  emailing: false,
  emailing_parameters: false,
  emailing_group: false,
  proposal_revisions: false,
  import_proposals: false,
  analytics_page: false,
  http_redirects: false,
  project_admin: false,
  developer_documentation: false,
  export_legacy_users: false,
  graphql_query_analytics: false,
  noindex_on_profiles: false,
  new_project_card: false,
  indexation: false,
  twilio: false,
  remind_user_account_confirmation: false,
  anonymous_questionnaire: false,
  login_cas: false,
  login_openid: false,
  public_api: false,
  sentry_log: false,
  versions: false,
  paper_vote: false,
  graphql_introspection: false,
  helpscout_beacon: false,
  api_sendinblue: false,
  proposal_sms_vote: false,
  organizations: false,
  moderation_comment: false,
  unstable__new_create_project: false,
  new_vote_step: false,
  mediator: false,
  online_help: false,
  full_proposal_card: false,
  multi_consultations: false,
}

type FeatureFlagsEnabled<T> = {
  [key in T as FeatureFlagType]?: boolean
}

export const useFeatureFlag = (flag: FeatureFlagType): boolean => {
  const data = useLazyLoadQuery<useFeatureFlagQuery>(query, {}, { fetchPolicy: 'store-only' })
  invariant(data.featureFlags !== undefined, 'The featureFlags are missing in Relay store.')

  const featureFlag = data.featureFlags.find(f => f.type === flag)
  invariant(featureFlag !== undefined, 'The featureFlag "%s" is missing in Relay store.', flag)
  return featureFlag ? featureFlag.enabled : false
}

export const useFeatureFlags = <T extends FeatureFlagType>(flags: FeatureFlagType[]): FeatureFlagsEnabled<T> => {
  const data = useLazyLoadQuery<useFeatureFlagQuery>(query, {}, { fetchPolicy: 'store-only' })
  invariant(data.featureFlags !== undefined, 'The featureFlags are missing in Relay store.')

  return flags.reduce<FeatureFlagsEnabled<T>>((acc, featureFlag) => {
    const currentFlag = data.featureFlags.find(f => f.type === featureFlag)

    if (currentFlag) {
      acc[featureFlag] = currentFlag.enabled
    } else {
      invariant(currentFlag !== undefined, 'The featureFlag "%s" is missing in Relay store.', featureFlag)
    }

    return acc
  }, {})
}

export const useAllFeatureFlags = (): FeatureFlags => {
  const data = useLazyLoadQuery<useFeatureFlagQuery>(query, {}, { fetchPolicy: 'store-only' })
  invariant(data.featureFlags !== undefined, 'The featureFlags are missing in Relay store.')

  return data.featureFlags.reduce<FeatureFlags>((acc, featureFlag) => {
    acc[featureFlag.type] = featureFlag.enabled
    return acc
  }, defaultFeatureFlags)
}

export default useFeatureFlag
