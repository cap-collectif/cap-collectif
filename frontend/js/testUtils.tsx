import * as React from 'react'
import ReactDOM from 'react-dom'
import { ThemeProvider } from 'styled-components'
import { Provider } from 'react-redux'
import { Provider as ReakitProvider } from 'reakit'
import { IntlProvider } from 'react-intl-redux'
import { RelayEnvironmentProvider } from 'react-relay'
import { CapUIProvider } from '@cap-collectif/ui'
import { BrowserRouter as Router } from 'react-router-dom'
import { theme } from '~/styles/theme'
import appStore from '~/stores/AppStore'
import type { FeatureFlagType } from '~relay/useFeatureFlagQuery.graphql'
import { FormProvider, useForm } from 'react-hook-form'

export const mockRandomValues = () => {
  global.Math.random = () => 0.5
}
export const enableFeatureFlags = (flags: Array<FeatureFlagType>) => {
  global.mockFeatureFlag.mockImplementation((flag: Array<FeatureFlagType>) => {
    if (flags.includes(flag)) {
      return true
    }

    return false
  })
}
export const disableFeatureFlags = () => {
  global.mockFeatureFlag.mockImplementation(() => false)
}
export const mockUrl = (url: string) => {
  // https://stackoverflow.com/questions/54021037/how-to-mock-window-location-href-with-jest-vuejs
  delete window.location
  // @ts-expect-error assign new URL(...) to window.location because property location is not writable
  window.location = new URL(url)
}
export const addsSupportForPortals = () => {
  // See: https://github.com/facebook/react/issues/11565
  // @ts-expect-error
  ReactDOM.createPortal = jest.fn(element => {
    return element
  })
}
export const clearSupportForPortals = () => {
  // @ts-expect-error
  ReactDOM.createPortal.mockClear()
}
type Props = {
  children: JSX.Element | JSX.Element[] | string
  store?: any
  useCapUIProvider?: boolean
}
export const MockProviders = ({ children, store = {}, useCapUIProvider = false }: Props) => {
  return (
    <Provider store={appStore(store)}>
      <Router>
        <IntlProvider>
          <ReakitProvider>
            {useCapUIProvider ? (
              <CapUIProvider>{children}</CapUIProvider>
            ) : (
              <ThemeProvider theme={theme}>{children}</ThemeProvider>
            )}
          </ReakitProvider>
        </IntlProvider>
      </Router>
    </Provider>
  )
}
type RelaySuspensFragmentTestProps = Props & {
  environment: any
}
export const RelaySuspensFragmentTest = ({
  children,
  environment,
  store = {},
  useCapUIProvider = false,
}: RelaySuspensFragmentTestProps) => {
  return (
    <MockProviders store={store} useCapUIProvider={useCapUIProvider}>
      <RelayEnvironmentProvider environment={environment}>
        <React.Suspense fallback="">{children}</React.Suspense>
      </RelayEnvironmentProvider>
    </MockProviders>
  )
}

export const FormWrapper = props => {
  const formMethods = useForm()

  return <FormProvider {...formMethods}>{props.children}</FormProvider>
}

export default MockProviders
