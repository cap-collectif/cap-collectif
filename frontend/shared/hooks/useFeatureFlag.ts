import { useLazyLoadQuery, graphql } from 'react-relay'
import invariant from '../utils/invariant'
import { useFeatureFlagQuery, FeatureFlagType } from '@relay/useFeatureFlagQuery.graphql'

const query = graphql`
  query useFeatureFlagQuery {
    featureFlags {
      type
      enabled
    }
  }
`

type FeatureFlagsEnabled<T> = {
  [key in T as FeatureFlagType]?: boolean
}

export const useFeatureFlag = (flag: FeatureFlagType): boolean => {
  const data = useLazyLoadQuery<useFeatureFlagQuery>(query, {}, { fetchPolicy: 'store-only' })
  invariant(data.featureFlags !== undefined, 'The featureFlags are missing in Relay store.')

  const featureFlag = data.featureFlags.find(f => f.type === flag)
  invariant(featureFlag !== undefined, 'The featureFlag "%s" is missing in Relay store.', flag)
  return featureFlag ? featureFlag.enabled : false
}

export const useFeatureFlags = <T extends FeatureFlagType>(flags: FeatureFlagType[]): FeatureFlagsEnabled<T> => {
  const data = useLazyLoadQuery<useFeatureFlagQuery>(query, {}, { fetchPolicy: 'store-only' })
  invariant(data.featureFlags !== undefined, 'The featureFlags are missing in Relay store.')

  return flags.reduce<FeatureFlagsEnabled<T>>((acc, featureFlag) => {
    const currentFlag = data.featureFlags.find(f => f.type === featureFlag)

    if (currentFlag) {
      acc[featureFlag] = currentFlag.enabled
    } else {
      invariant(currentFlag !== undefined, 'The featureFlag "%s" is missing in Relay store.', featureFlag)
    }

    return acc
  }, {})
}

export default useFeatureFlag
