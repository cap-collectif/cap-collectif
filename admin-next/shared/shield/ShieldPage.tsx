import * as React from 'react'
import { useIntl } from 'react-intl'
import { graphql, useLazyLoadQuery } from 'react-relay'
import { LoginButtonQueryWrapper } from '@shared/login/LoginButton'
import LoginBox from '@shared/login/LoginBox'
import { RegistrationButtonQueryWrapper } from '@shared/register/RegistrationButton'
import { ShieldPageQuery, ShieldPageQuery$data } from '@relay/ShieldPageQuery.graphql'
import useFeatureFlag from '@shared/hooks/useFeatureFlag'
import LoginFormWrapper from '@shared/login/LoginFormWrapper'
import { useFormContext } from 'react-hook-form'
import { Button } from '@cap-collectif/ui'

type Props = {
  loginWithOpenId: boolean
  query: ShieldPageQuery$data
}

const QUERY = graphql`
  query ShieldPageQuery {
    ...LoginBox_query
    oauth2sso: ssoConfigurations(ssoType: OAUTH2) {
      edges {
        node {
          enabled
        }
      }
    }
  }
`

const SimpleShieldForm = ({ query }: { query: ShieldPageQuery$data }) => {
  const intl = useIntl()
  const {
    formState: { isSubmitting },
  } = useFormContext()

  return (
    <>
      <LoginBox query={query} />
      <Button
        variant="primary"
        id="confirm-login"
        type="submit"
        isLoading={isSubmitting}
        mt={4}
        width="100%"
        justifyContent="center"
      >
        {intl.formatMessage({ id: isSubmitting ? 'global.loading' : 'global.login_me' })}
      </Button>
    </>
  )
}

export const ShieldBody = ({ query, loginWithOpenId }: Props) => {
  const byPassAuth = useFeatureFlag('sso_by_pass_auth')
  const showRegistration = useFeatureFlag('registration')

  if (showRegistration && !loginWithOpenId) {
    return (
      <>
        <LoginButtonQueryWrapper justifyContent="center" width="100%" mb={2} />
        <RegistrationButtonQueryWrapper justifyContent="center" width="100%" />
      </>
    )
  }

  if (byPassAuth) return <LoginBox query={query} />

  return (
    <LoginFormWrapper>
      <SimpleShieldForm query={query} />
    </LoginFormWrapper>
  )
}

export const ShieldPage = () => {
  const query = useLazyLoadQuery<ShieldPageQuery>(QUERY, {})

  return (
    <div id="shield-agent" className="bg-white col-md-4 col-md-offset-4 panel panel-default">
      <div className="panel-body">
        <ShieldBody query={query} loginWithOpenId={query.oauth2sso.edges.some(({ node }) => node.enabled)} />
      </div>
    </div>
  )
}

export default ShieldPage
