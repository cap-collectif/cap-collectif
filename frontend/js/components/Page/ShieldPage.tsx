import * as React from 'react'
import { useIntl } from 'react-intl'
import { graphql, useLazyLoadQuery } from 'react-relay'
import LoginButton from '../User/Login/LoginButton'
import LoginBox from '@shared/login/LoginBox'
import RegistrationButton from '../User/Registration/RegistrationButton'
import RegistrationModal from '~/components/User/Registration/RegistrationModal'
import { ShieldPageQuery, ShieldPageQuery$data } from '~relay/ShieldPageQuery.graphql'
import useFeatureFlag from '@shared/hooks/useFeatureFlag'
import LoginFormWrapper from '@shared/login/LoginFormWrapper'
import { useFormContext } from 'react-hook-form'
import { Button } from '@cap-collectif/ui'

type Props = {
  loginWithOpenId: boolean
  query: ShieldPageQuery$data
}

const QUERY = graphql`
  query ShieldPageQuery {
    ...RegistrationModal_query
    ...LoginBox_query
    oauth2sso: ssoConfigurations(ssoType: OAUTH2) {
      edges {
        node {
          enabled
        }
      }
    }
  }
`

const SimpleShieldForm = ({ query }: { query: ShieldPageQuery$data }) => {
  const intl = useIntl()
  const {
    formState: { isSubmitting },
  } = useFormContext()

  return (
    <>
      <LoginBox query={query} />
      <Button
        variant="primary"
        id="confirm-login"
        type="submit"
        isLoading={isSubmitting}
        mt={4}
        width="100%"
        justifyContent="center"
      >
        {intl.formatMessage({ id: isSubmitting ? 'global.loading' : 'global.login_me' })}
      </Button>
    </>
  )
}

export const ShieldBody = ({ query, loginWithOpenId }: Props) => {
  const byPassAuth = useFeatureFlag('sso_by_pass_auth')
  const showRegistration = useFeatureFlag('registration')

  if (showRegistration && !loginWithOpenId) {
    return (
      <>
        <LoginButton justifyContent="center" width="100%" mb={2} />
        <RegistrationButton className="btn-block" />
      </>
    )
  }

  if (byPassAuth) return <LoginBox query={query} />

  return (
    <LoginFormWrapper>
      <SimpleShieldForm query={query} />
    </LoginFormWrapper>
  )
}

export const ShieldPage = ({ locale }: { locale: string }) => {
  const query = useLazyLoadQuery<ShieldPageQuery>(QUERY, {})

  return (
    <div id="shield-agent" className="bg-white col-md-4 col-md-offset-4 panel panel-default">
      <RegistrationModal query={query} locale={locale} />
      <div className="panel-body">
        <ShieldBody query={query} loginWithOpenId={query.oauth2sso.edges.some(({ node }) => node.enabled)} />
      </div>
    </div>
  )
}

export default ShieldPage
