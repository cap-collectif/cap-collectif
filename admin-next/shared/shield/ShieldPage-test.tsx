/* eslint-env jest */
import { render } from '@testing-library/react'
import { graphql, useLazyLoadQuery } from 'react-relay'
import { createMockEnvironment, MockPayloadGenerator } from 'relay-test-utils'
import type { ShieldPageTestQuery } from '@relay/ShieldPageTestQuery.graphql'
import { ShieldBody } from './ShieldPage'

import useFeatureFlag from '@shared/hooks/useFeatureFlag'
import { RelaySuspensFragmentTest } from 'tests/testUtils'

jest.mock('@shared/hooks/useFeatureFlag')

describe('<ShieldPage />', () => {
  let environment
  let TestComponent
  const defaultMockResolvers = {
    Query: () => ({
      oauth2sso: {
        edges: [
          {
            node: {
              enabled: true,
            },
          },
          {
            node: {
              enabled: false,
            },
          },
        ],
      },
    }),
  }
  const query = graphql`
    query ShieldPageTestQuery @relay_test_operation {
      ...LoginBox_query
      oauth2sso: ssoConfigurations(ssoType: OAUTH2) {
        edges {
          node {
            enabled
          }
        }
      }
    }
  `

  beforeEach(() => {
    environment = createMockEnvironment()
    environment.mock.queueOperationResolver(operation => MockPayloadGenerator.generate(operation, defaultMockResolvers))

    const TestRenderer = props => {
      const data = useLazyLoadQuery<ShieldPageTestQuery>(query, {})
      if (!data) return null
      return <ShieldBody query={data} {...props} />
    }

    TestComponent = props => (
      <RelaySuspensFragmentTest environment={environment}>
        <TestRenderer {...props} />
      </RelaySuspensFragmentTest>
    )
  })
  it('renders with registration enabled', () => {
    // @ts-ignore jest
    useFeatureFlag.mockImplementation(e => e === 'registration')
    const { asFragment } = render(<TestComponent />)
    expect(asFragment()).toMatchSnapshot()
  })
  it('renders with registration disabled', () => {
    // @ts-ignore jest
    useFeatureFlag.mockReturnValue(false)
    const { asFragment } = render(<TestComponent />)
    expect(asFragment()).toMatchSnapshot()
  })
  it('renders with registration disabled and classical auth by pass enabled', () => {
    // @ts-ignore jest
    useFeatureFlag.mockImplementation(e => e === 'sso_by_pass_auth')
    const { asFragment } = render(<TestComponent />)
    expect(asFragment()).toMatchSnapshot()
  })
})
