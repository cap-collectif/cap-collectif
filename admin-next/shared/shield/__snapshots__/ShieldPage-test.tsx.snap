// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ShieldPage /> renders with registration disabled 1`] = `
[
  <div
    className="sc-bcXGCL gHoLKQ toasts-container toasts-container--top"
  />,
  <form
    id="login-form"
    onSubmit={[Function]}
  >
    <div>
      <div
        className="sc-bcXGCL iAhoTX"
      >
        <div
          className=" ql-editor wysiwyg-render"
          dangerouslySetInnerHTML={
            {
              "__html": "<mock-value-for-field-"value">",
            }
          }
        />
      </div>
      <div
        className="sc-bcXGCL fVgOgb cap-form-control"
      >
        <label
          className="sc-bcXGCL fTPHCc cap-form-label"
          htmlFor=":r0:"
          style={
            {
              "marginLeft": 0,
              "marginTop": 0,
            }
          }
        >
          <span
            className="sc-bcXGCL dQChsN cap-text"
            style={
              {
                "marginLeft": 0,
                "marginTop": 0,
              }
            }
          >
            global.email
          </span>
        </label>
        <input
          aria-required={true}
          autoComplete="email"
          autoFocus={true}
          className="sc-bcXGCL sc-dwnPIk egzrup eCNeCU cap-input"
          disabled={false}
          id=":r0:"
          name="username"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="email.placeholder"
          readOnly={false}
          required={true}
          type="email"
        />
      </div>
      <div
        className="sc-bcXGCL fVgOgb cap-form-control"
      >
        <label
          className="sc-bcXGCL fTPHCc cap-form-label"
          htmlFor=":r1:"
          style={
            {
              "marginLeft": 0,
              "marginTop": 0,
            }
          }
        >
          <span
            className="sc-bcXGCL dQChsN cap-text"
            style={
              {
                "marginLeft": 0,
                "marginTop": 0,
              }
            }
          >
            global.password
          </span>
        </label>
        <input
          aria-required={true}
          autoComplete="off"
          className="sc-bcXGCL sc-dwnPIk egzrup eCNeCU cap-input"
          disabled={false}
          id=":r1:"
          name="password"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          readOnly={false}
          required={true}
          type="password"
        />
      </div>
      <a
        className="sc-bcXGCL gudgoe cap-link"
        href="/resetting/request"
      >
        global.forgot_password
      </a>
      <div
        className="sc-bcXGCL gvdeaM"
      >
        <div
          className=" ql-editor wysiwyg-render"
          dangerouslySetInnerHTML={
            {
              "__html": "<mock-value-for-field-"value">",
            }
          }
        />
      </div>
    </div>
    <button
      className="sc-bcXGCL sc-gGvGpy hRwdLZ jrAhSN cap-button"
      id="confirm-login"
      type="submit"
    >
      global.login_me
    </button>
  </form>,
]
`;

exports[`<ShieldPage /> renders with registration disabled and classical auth by pass enabled 1`] = `
[
  <div
    className="sc-bcXGCL gHoLKQ toasts-container toasts-container--top"
  />,
  <div>
    <div
      className="sc-bcXGCL iAhoTX"
    >
      <div
        className=" ql-editor wysiwyg-render"
        dangerouslySetInnerHTML={
          {
            "__html": "<mock-value-for-field-"value">",
          }
        }
      />
    </div>
    <div
      className="sc-bcXGCL gvdeaM"
    >
      <div
        className=" ql-editor wysiwyg-render"
        dangerouslySetInnerHTML={
          {
            "__html": "<mock-value-for-field-"value">",
          }
        }
      />
    </div>
  </div>,
]
`;

exports[`<ShieldPage /> renders with registration enabled 1`] = `
<div
  className="sc-bcXGCL gHoLKQ toasts-container toasts-container--top"
/>
`;
